ssm/observations.py

```diff
@@ -892,64 +893,146 @@ def log_likelihoods(self, data, input, mask, tag=None):
 
         return np.row_stack((ll_init, ll_ar))
 
-    def m_step(self, expectations, datas, inputs, masks, tags, J0=None, h0=None):
+    def m_step(self, expectations, datas, inputs, masks, tags,
+               J0=None, h0=None, continuous_expectations=None, **kwargs):
+        """Compute M-step for Gaussian Auto Regressive Observations.
+
+        If `continuous_expectations` is not None, this function will
+        compute an exact M-step using the expected sufficient statistics for the
+        continuous states. In this case, we ignore the prior provided by (J0, h0),
+        because the calculation is exact. `continuous_expectations` should be a tuple of
+        (Ex, Ey, ExxT, ExyT, EyyT). 
+
+        If `continuous_expectations` is None, we use `datas` and `expectations,
+        and (optionally) the prior given by (J0, h0). In this case, we estimate the sufficient
+        statistics using `datas,` which is typically a single sample of the continuous
+        states from the posterior distribution. 
+        """
         K, D, M, lags = self.K, self.D, self.M, self.lags
 
-        # Collect all the data
+        As = np.zeros((K, D, D * lags + M))
+        bs = np.zeros((K, D))
+        Sigmas = np.zeros((K, D, D))
         xs, ys, Ezs = [], [], []
         for (Ez, _, _), data, input, mask, tag in zip(expectations, datas, inputs, masks, tags):
```

Comment:
We only need to execute this block if the `continuous_expectations` are not given.  It's memory intensive so let's only do it if necessary.

Reply:
Postponing this for a later date.